workers-scaling/README.md

```diff
@@ -0,0 +1,88 @@
+by making sure that the workers are not so greedy and reserve only one task ahead
+(`--prefetch-multiplier=1`, default is 4).
+
+Later, we also [set up 2 Celery queues, for long and short running tasks](https://github.com/packit/packit-service/commit/617c76ef77ef37bc220b029bc49326d3841c54c0)
+because that's what [Celery User Guide suggests](https://docs.celeryq.dev/en/stable/userguide/optimizing.html#optimizing-prefetch-limit).
+
+## Concurrency
+
+The same [commit](<(https://github.com/packit/packit-service/commit/5d199cfee54dafd9a5cd5dadc1086e15d78598e5)>)
+that set `--prefetch-multiplier=1` also set `--concurrency=1`
+(number of concurrent worker processes/threads executing tasks)
+even it's probably had no effect because [the default is "number of CPUs/cores on the host"](https://docs.celeryq.dev/en/stable/userguide/configuration.html#std-setting-worker_concurrency)
+and the workers have no more than one core.
+We definitely don't want to change the prefetch multiplier because that might
+cause [p-s#375](https://github.com/packit/packit-service/issues/375)
+for no significant benefit, but having more concurrent worker threads
+should hopefully be safe.
+
+Whether it's really safe depends on how Celery
+fetches messages for the worker's threads. If it waits for all threads
+to finish their tasks before fetching a new batch of messages (times prefetch-multiplier)
+for all threads then we might still hit [p-s#375](https://github.com/packit/packit-service/issues/375)
+if we have more workers serving the short-running-tasks queue.
+But if Celery fetches a new message(s) for a thread once it finishes a task
+without waiting for other threads then we should be fine (as long as prefetch-multiplier=1).
+I checked the [Celery source code](https://github.com/celery/celery)
+but couldn't find an answer for how/when it fetches messages for the threads
+(whether individually or for all at once). Someone's up for a challenge?;)
+But if we had only one worker (with more threads) serving the
+short-running-tasks queue then we should be fine anyway.
+
+#### Processes vs. threads aka. [Celery execution pool](https://www.distributedpython.com/2018/10/26/celery-execution-pool)
+
+- **Processes** (`prefork` pool) -
+  are good when tasks are CPU bound. The number of processes should not exceed
+  the number of CPUs on host.
+- **Threads** ([eventlet](https://docs.celeryq.dev/en/latest/userguide/concurrency/eventlet.html)/[gevent](https://www.gevent.org) pool) -
+  when tasks are I/O or network bound. The number of greenlets (green threads) is
+  unrelated to the number of CPUs on host and can go even to thousands.
+
+#### Autoscaling
+
+Regarding the `--concurrency` value, we even don't have to set a static number
+if we use [worker --autoscale](https://docs.celeryq.dev/en/latest/userguide/workers.html#autoscaling)
+which dynamically resizes the pool based on load.
```

Comment:
Also, what about a situation where we have a huge queue of inexpensive tasks (if I get that right, this will not show up in a higher load).

Reply:
The load here most likely doesn't mean CPU load, because Celery has no clue about such metric, but number of requests in the message queue - I checked the code, and Celery just [compares](https://github.com/celery/celery/blob/aa9fd8a6c06e69c7eda2a59866c3d84622c85d20/celery/worker/autoscale.py#L85) whether all threads have [something to do](https://github.com/celery/celery/blob/aa9fd8a6c06e69c7eda2a59866c3d84622c85d20/celery/worker/autoscale.py#L150) and if so and we haven't scaled to max threads, add more threads.

I updated the text.

Reply:
Nice, thanks for the explanation!